### ISSUE TITLE
Emacs locks twice for a few seconds when saving go file 

### ISSUE THREAD

**Issue by rodrigovalin** [dominikh/go-mode.el#273]
This thing has been driving me crazy for the last few days, without being able to find the cause. Emacs is freezing for around 4 seconds, two times, every time I save a go file. It just hangs for a bit. I've removed every other package (and every binary I could find) and came to the conclussion that this is go-mode doing some directory traversal and taking a while to do so. I straced the emacs process when it freezes and I get something like:

```
readlinkat(AT_FDCWD, "/home", 0x7ffdc18ed500, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin", 0x7ffdc18ed890, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace", 0x7ffdc18edc20, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go", 0x7ffdc18edfb0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src", 0x7ffdc18ee340, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com", 0x7ffdc18ee6d0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/10gen", 0x7ffdc18eea60, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage", 0x7ffdc18eedf0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg", 0x7ffdc18ef180, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller", 0x7ffdc18ef510, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller/om", 0x7ffdc18ef8a0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller/om/omclient.go", 0x7ffdc18efc30, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home", 0x7ffdc18ed2c0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin", 0x7ffdc18ed650, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace", 0x7ffdc18ed9e0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go", 0x7ffdc18edd70, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src", 0x7ffdc18ee100, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com", 0x7ffdc18ee490, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/10gen", 0x7ffdc18ee820, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage", 0x7ffdc18eebb0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg", 0x7ffdc18eef40, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller", 0x7ffdc18ef2d0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller/om", 0x7ffdc18ef660, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller/om/omclient.go", 0x7ffdc18ef9f0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home", 0x7ffdc18ed2c0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin", 0x7ffdc18ed650, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace", 0x7ffdc18ed9e0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go", 0x7ffdc18edd70, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src", 0x7ffdc18ee100, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com", 0x7ffdc18ee490, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/10gen", 0x7ffdc18ee820, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage", 0x7ffdc18eebb0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg", 0x7ffdc18eef40, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller", 0x7ffdc18ef2d0, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller/om", 0x7ffdc18ef660, 1024) = -1 EINVAL (Invalid argument)
readlinkat(AT_FDCWD, "/home/rvalin/workspace/go/src/github.com/someorg/somepackage/pkg/controller/om/process_test.go", 0x7ffdc18ef9f0, 1024) = -1 EINVAL (Invalid argument
```

This repeats probably hundreds of time, during which, Emacs is completely frozen.

My question is, how can I debug go-mode.el myself to see where is this happening? Do you have any idea who might be responsible of this?

**Comment by dominikh** [comment]
> I've removed every other package (and every binary I could find) and came to the conclussion that this is go-mode doing some directory traversal and taking a while to do so

To make sure that you're dealing with a 100% clean Emacs install, run it with the `-Q` flag, to skip loading of any configuration, user or system alike. Then you can load just go-mode.el and be sure that nothing else is running.

Also, strace Emacs while saving a buffer that isn't in go-mode so you can compare the traces. Otherwise there's no telling if what you're seeing is actually symptomatic.

> Emacs is freezing for around 4 seconds, two times, every time I save a go file

Can you expand on how it freezes twice? What happens between the two freezes that you can tell it unfroze? 

As for debugging go-mode.el, you can look into https://www.gnu.org/software/emacs/manual/html_node/elisp/Using-Edebug.html – alas, I'm not sure go-mode is at fault here. The only thing that hooks into saving buffers is the gofmt hook, and that just runs gofmt once before saving. But of course you could instrument it to confirm that.

**Comment by rodrigovalin** [comment]
Thanks a lot for the detailed instructions on how to proceed. I'll do all the required investigation and come back with more data.

**Comment by rodrigovalin** [comment]
1. I started stripping down my emacs configuration (removing everything from custom.el) but no changes.
2. I started with a clean install, then installed prelude and activated the `prelude-go` module (I know that I'm going to a fully fledged distro now and that it is difficult to debug at this point)
3. As soon as I visited a `go` source file Emacs will start freezing at different intervals and the output in my first comment will start appearing, _hundreds_ of times, making Emacs completely unusable
4. I tried with go1.12 and go1.11 -- downloaded from official tar.gz distribution
5. I also tried with go1.11 packaged for Ubuntu (https://github.com/golang/go/wiki/Ubuntu)

No changes! I could not expect that any of the installed packages (dependencies of `projectile-go` module) would be so buggy so I checked my `$GOPATH/bin` directory for things that could be messing with my Emacs install (maybe an old & buggy `gocode`?)

I removed everything from `$GOPATH/bin` and voila! no more unresponsive Emacs. And no more, jumping to definition, auto complete, linter, nothing. I started `go get`ing stuff as I needed them and came to the conclusion that I actually needed: `gocode`, `godef`. For now, at least to have something operational to finish my week.

I don't have the complete list of programs (I'm writing this from my personal computer) that I removed, but I'm sure I was using an old version of something that would get stuck or not act how it was supposed to.

Closing this ticket now, as the issue appears to be a local problem with my installation.
